cms/wizards/wizard_pool.py

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+
+from django.utils.translation import ugettext as _
+
+from cms.utils.django_load import load
+
+from .wizard_base import Wizard
+
+
+class AlreadyRegisteredException(Exception):
+    pass
+
+
+class WizardPool(object):
+    _entries = {}
+    _discovered = False
+
+    def __init__(self):
+        self._reset()
+
+    # PRIVATE METHODS -----------------
+
+    def _discover(self):
+        if not self._discovered:
+            load('cms_wizards')
+            self._discovered = True
+
+    def _clear(self):
+        """Simply empties the pool but does not clear the discovered flag."""
+        self._entries = {}
+
+    def _reset(self):
+        """Clears the wizard pool and clears the discovered flag."""
+        self._clear()
+        self._discovered = False
+
+    # PUBLIC METHODS ------------------
+
+    @property
+    def discovered(self):
+        """
+        A public getter for the private property _discovered. Note, there is no
+        public setter.
+        """
+        return self._discovered
+
+    def is_registered(self, entry):
+        """
+        Returns True if the provided entry is registered.
+
+        NOTE: This method triggers pool discovery.
+        """
+        self._discover()
+        return entry.id in self._entries
+
+    def register(self, entry):
+        """
+        Registers the provided «entry».
+
+        Raises AlreadyRegisteredException if the entry is already registered.
+
+        NOTE: This method triggers pool discovery.
+        """
+        assert isinstance(entry, Wizard), u"entry must be an instance of Wizard"
+        if self.is_registered(entry):
```

Comment:
I'm not sure about any infinite loop possibilities, but I agree it might create unexpected results. To counter any of this, I've publicly exposed the property `discovered`. It can be used to first check if the pool is populated before using any of these other methods. If the pool hasn't been populated, then it is already nonsensical to use any of those methods. The outcome of each is already known.

However to put any further concerns to rest, I'll create a kwarg on is_registered() that will not trigger discovery.